[clang] Implement CWG1878 "`operator auto` template"

Endilll opened the pull request:
C++14 introduced deduced return type for regular functions, but shortly after [CWG1878](https://wg21.link/cwg1878) was filed and resolved to disallow deduced return types in conversion function templates. So this patch diagnoses such usage of deduced return type in C++14 mode onwards.

Fixes #51776

Endilll committed:
[clang] Implement CWG1878 "`operator auto` template"

C++14 introduced deduced return type for regular functions, but shortly after [CWG1878](https://wg21.link/cwg1878) was filed and resolved to disallow deduced return types in conversion function templates. So this patch diagnoses such usage of deduced return type in C++14 mode onwards.

Endilll committed:
Run clang-format

cor3ntin commented on clang/test/SemaCXX/deduced-return-type-cxx14.cpp:
```diff
@@ -686,3 +686,19 @@ auto f(auto x) { // cxx14-error {{'auto' not allowed in function prototype}}
+
+struct DeducedTargetTypeOfConversionFunction {
+  operator auto() const { return char(); }
+  operator const auto() const { return float(); }
```
can you add tests with references?

Endilll replied:
Done

cor3ntin commented on clang/lib/Sema/SemaDeclCXX.cpp:
```diff
@@ -11322,9 +11322,22 @@ Decl *Sema::ActOnConversionDeclarator(CXXConversionDecl *Conversion) {
+      Diag(Conversion->getTypeSpecStartLoc(), diag::err_auto_not_allowed)
+          << Conversion->getTypeSourceInfo()
+                 ->getTypeLoc()
+                 .getContainedAutoTypeLoc()
```
I think this is a slightly bigger hammer than it needs to be.
I think we should just use `->getTypeLoc().getSourceRange()` (yes, it would be the source range of the type + qualifiers, but i think that's as good, if not better

Endilll replied:
I went this way mostly because I didn't want `const` qualifier on the member function to be highlighted, as it's not relevant. But it's getting a bit worse than this if we go your way:
```
    4 |   operator const auto() const { return 1.2; }     // error: conversion function template
      |                  ^~~~~~~~~~~~
    6 |   operator decltype(auto)() const { return 1.2; }     // error: conversion function template
      |            ^~~~~~~~~~~~~~~~~~~~~~
```

cor3ntin replied:
Just the  return type.
There is even simpler actually

```cpp
getReturnTypeLoc(Conversion).getSourceRange()
```

Endilll replied:
Nice! I updated the PR.

cor3ntin commented on clang/lib/Sema/SemaDeclCXX.cpp:
```diff
@@ -11322,9 +11322,22 @@ Decl *Sema::ActOnConversionDeclarator(CXXConversionDecl *Conversion) {
+                 ->getTypeLoc()
+                 .getContainedAutoTypeLoc()
+                 .getSourceRange()
+          << llvm::to_underlying(Conversion->getConversionType()
+                                     ->getAs<AutoType>()
+                                     ->getKeyword())
```
You do not have to change it here, but this is kind of nasty.
`<< Conversion->getConversionType()` should just work, so this is an artifact of a weirdly constructed diagnostic.
I wonder if we should adapt the `err_auto_not_allowed` diagnostic.

We do this weird dance in 3 places
I also will not properly diagnose constrained return types

ie ```cpp
template <typename T>
concept C = true;

struct S {
    operator C auto();
};```


Writing that iI think we should bite the bullet and change it

```cpp

def err_auto_not_allowed : Error<
  "%select{"use of "
  "%select{class template|function template|variable template|alias template|"
  "template template parameter|concept|template}3 %4 requires template "
  "arguments; argument deduction|%2}0 not allowed "
  "%select{in function prototype"
  "|in non-static struct member|in struct member"
  "|in non-static union member|in union member"
  "|in non-static class member|in interface member"
  "|in exception declaration|in template parameter until C++17|in block literal"
  "|in template argument|in typedef|in type alias|in function return type"
  "|in conversion function type|here|in lambda parameter"
  "|in type allocated by 'new'|in K&R-style function parameter"
  "|in template parameter|in friend declaration|in function prototype that is "
  "not a function declaration|in requires expression parameter"
  "|in array declaration"
  "|in declaration of conversion function template}1">;
```
Should work, we would have to change the 3 uses of err_auto_not_allowed

Can you add tests for constrained auto?

Endilll committed:
Simplify extracting source range info for return type

Endilll committed:
Address Corentin's feedback

Add tests with constrained auto and references, adapt diagnostic emitter accordingly.

Endilll committed:
Merge remote-tracking branch 'upstream/main' into cwg1878

Endilll committed:
Add `auto *` test and simplify diag emitting logic

cor3ntin commented on clang/lib/Sema/SemaDeclCXX.cpp:
```diff
@@ -11321,9 +11321,20 @@ Decl *Sema::ActOnConversionDeclarator(CXXConversionDecl *Conversion) {
+      Diag(Conversion->getTypeSpecStartLoc(), diag::err_auto_not_allowed)
+          << getReturnTypeLoc(Conversion).getSourceRange()
+          << llvm::to_underlying(ConvType->getAs<AutoType>()->getKeyword())
+          << /* in declaration of conversion function template */ 24;
```
```suggestion
          << /* in declaration of conversion function template= */ 24;
```

cor3ntin reviewed (approved):
LGTM, thanks
Feel free to refactor `err_auto_not_allowed` in a separate PR

Endilll committed:
Apply Corentin's fix

shafik commented on clang/lib/Sema/SemaDeclCXX.cpp:
```diff
@@ -11321,9 +11321,20 @@ Decl *Sema::ActOnConversionDeclarator(CXXConversionDecl *Conversion) {
+      Diag(Conversion->getTypeSpecStartLoc(), diag::err_auto_not_allowed)
+          << getReturnTypeLoc(Conversion).getSourceRange()
+          << llvm::to_underlying(ConvType->getAs<AutoType>()->getKeyword())
+          << /* in declaration of conversion function template= */ 24;
```
`warn_cconv_unsupported` uses `CallingConventionIgnoredReason` enum to specify which one.

Not sure if this is something we want to use more often or not.
